File: backend/src/main/scala/io/github/nwtgck/gh_card/infra/RedisGitHubRepositoryJsonCacheRepository.scala
package io.github.nwtgck.gh_card.infra

import com.redis.RedisClientPool
import io.github.nwtgck.gh_card.domain.GitHubRepositoryJsonCacheRepository

class RedisGitHubRepositoryJsonCacheRepository(redisClientPool: RedisClientPool, ttl: Int) extends GitHubRepositoryJsonCacheRepository {
  val keyPrefix = "repos-json"

  def cache(repoName: String, json: String): Unit = {
    val key = s"${keyPrefix}/${repoName}"
    redisClientPool.withClient{ client =>
      client.set(key, json)
      client.expire(key, ttl)
    }
  }

  def get(repoName: String): Option[String] = {
    val key = s"${keyPrefix}/${repoName}"
    redisClientPool.withClient{client =>
      client.get(key)
    }
  }
}

File: backend/src/infra/RedisGitHubRepositoryJsonCacheRepository.ts
import * as redis from 'redis';
import {promisify} from 'util';

import {GitHubRepositoryJsonCacheRepository} from '../domain/GitHubRepositoryJsonCacheRepository';

export class RedisGitHubRepositoryJsonCacheRepository implements GitHubRepositoryJsonCacheRepository {
  private keyPrefix = "repos-json";
  private redisGetAsync: (key: string) => Promise<string>;

  constructor(private readonly redisClient: redis.RedisClient, private readonly ttlSec: number) {
    this.redisGetAsync = promisify(this.redisClient.get).bind(this.redisClient);
  }

  cache(repoName: string, json: string): void {
    this.redisClient.setex(`${this.keyPrefix}/${repoName}`, this.ttlSec, json);
  }

  async get(repoName: string): Promise<string | undefined> {
    const jsonStr: string | null = await this.redisGetAsync(`${this.keyPrefix}/${repoName}`);
    return jsonStr ?? undefined;
  }
}
